[naga wgsl-in] Add support for unsigned types when calling textureLoad with the level parameter.

ygdrasil-io opened the pull request:
**Connections**
Fix https://github.com/gfx-rs/wgpu/issues/5433

**Description**
In WSGL, the Naga frontend returns an error when using a u32 as the level parameter in textureLoad, although this is acceptable according to the specification at https://www.w3.org/TR/WGSL/#textureload.

**Testing**
This issue was tested by updating the image.wsgl file, which is executed as part of the unit test suite. The test includes a call to textureLoad with a u32 parameter.

<!--
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're no bothering us!
-->

**Checklist**

- [X] Run `cargo fmt`.
- [X] Run `taplo format`.
- [X] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
- [X] Run `cargo xtask test` to run tests.
- [X] Add change to `CHANGELOG.md`. See simple instructions inside file.

ygdrasil-io committed:
Refactor image level index type validation logic.

Improve the validation logic for image level indices by matching against specific scalar types (Sint or Uint). This change ensures accurate error handling for invalid index types.

ygdrasil-io committed:
Run cargo fmt

ygdrasil-io committed:
Update CHANGELOG with new capabilities from RP

ygdrasil-io commented:
FYI To test I used `cargo test --all-features --workspace` because `cargo xtask test` return 
```
cargo xtask test
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.04s
     Running `xtask/target/debug/xtask test`
[INFO  xtask::test] Generating .gpuconfig file based on gpus on the system
   Compiling ...
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 6.85s
     Running `target/debug/wgpu-info --json -o .gpuconfig`
[INFO  xtask::test] Found 1 gpu
[INFO  xtask::test] Running cargo tests
error: no such command: `nextest`

        Did you mean `test`?

        View all installed commands with `cargo --list`
        Find a package to install `nextest` with `cargo search cargo-nextest`
Error: Tests failed

Caused by:
    command exited with non-zero code `cargo nextest run --benches --tests --no-fail-fast --all-features --retries 0`: 101
```

ErichDonGubler commented:
@ygdrasil-io: Sounds like you don't have `cargo-nextest` installed; you can find instructions to do so [here](https://nexte.st/#quick-start). Our diagnostic ought to be better, though.

ygdrasil-io commented:
The texelFetch function in GLSL does not support unsigned integers. Should we cast the value to an integer?
https://registry.khronos.org/OpenGL-Refpages/gl4/html/texelFetch.xhtml

ygdrasil-io committed:
Handle Uint to SInt casting for GLSL texture operations.

ygdrasil-io commented:
@cwfitzgerald suggested doing casting on matrix, so I implemented it with [commit 546497b](https://github.com/gfx-rs/wgpu/pull/7058/commits/546497bbe3cff6ae42341b3dfe6c96433dc90c5c).

cwfitzgerald commented on naga/tests/out/wgsl/image.wgsl:
```diff
@@ -48,6 +48,7 @@ fn main(@builtin(local_invocation_id) local_id: vec3<u32>) {
     let dim = textureDimensions(image_storage_src);
     let itc = (vec2<i32>((dim * local_id.xy)) % vec2<i32>(10i, 20i));
     let value1_ = textureLoad(image_mipmapped_src, itc, i32(local_id.z));
+    let value1_2_ = textureLoad(image_mipmapped_src, itc, u32(local_id.z));
```
Could we call out what's explicitly happening with a comment

cwfitzgerald replied:
Whoops this is the out file, I'll check what I meant again when I re-review this

ygdrasil-io replied:
I have modified the WSGL shader source file with a comment.

cwfitzgerald commented on naga/tests/out/hlsl/image.hlsl:
```diff
@@ -37,6 +37,7 @@ void main(uint3 local_id : SV_GroupThreadID)
     uint2 dim = NagaRWDimensions2D(image_storage_src);
     int2 itc = (int2((dim * local_id.xy)) % int2(10, 20));
     uint4 value1_ = image_mipmapped_src.Load(int3(itc, int(local_id.z)));
+    uint4 value1_2_ = image_mipmapped_src.Load(int3(itc, uint(local_id.z)));
```
I feel like there's an implicit conversion here, and we should also wrap the argument in `int` as well, like we do glsl.

ygdrasil-io replied:
This is done. I also did some refactoring on `wgpu/naga/src/back/hlsl/writer.rs` to break up the monolithic function `write_expr`.

cwfitzgerald reviewed (changes requested):
Nice, looks generally good, a few comments

ygdrasil-io committed:
Add comment about u32 support of textureLoad on WSGL test file

ygdrasil-io committed:
Refactor HLSL image load handling into a separate function

Moved HLSL image load logic into a dedicated `write_image_load` function to improve modularity and readability. Additionally, added explicit type casting for certain scalar values to ensure correctness in generated HLSL code. Updated tests to reflect these changes and added related utility updates.

ygdrasil-io committed:
Refactor HLSL writer function parameters.

Update parameter types in `write_expr` for consistency and add `#[allow(clippy::too_many_arguments)]` to suppress linter warnings. This improves readability and maintains compatibility with existing code conventions.

ygdrasil-io committed:
erratum

ygdrasil-io committed:
fix clippy

ygdrasil-io committed:
cargo fmt

jimblandy commented on naga/src/back/hlsl/help.rs:
```diff
@@ -380,7 +380,7 @@ impl<W: Write> super::Writer<'_, W> {
         constructor: WrappedConstructor,
     ) -> BackendResult {
-        let name = crate::TypeInner::hlsl_type_id(constructor.ty, module.to_ctx(), &self.names)?;
+        let name = TypeInner::hlsl_type_id(constructor.ty, module.to_ctx(), &self.names)?;
```
Can we please keep changes like this out of the PR? It makes the change much harder to review, for me at least.

cwfitzgerald commented on naga/tests/in/image.wgsl:
```diff
@@ -21,6 +21,8 @@ fn main(@builtin(local_invocation_id) local_id: vec3<u32>) {
     let itc = vec2<i32>(dim * local_id.xy) % vec2<i32>(10, 20);
     // loads with ivec2 coords.
     let value1 = textureLoad(image_mipmapped_src, itc, i32(local_id.z));
+    // doing the same thing as the line above, but with u32, as textureLoad must also support unsigned integers.
+    let value1_2 = textureLoad(image_mipmapped_src, itc, u32(local_id.z));
```
Nice!

cwfitzgerald reviewed (changes requested):
I think we're g2g after jimb's comment is addressed.

ygdrasil-io committed:
remove TypeInner import on help.rs

ygdrasil-io commented:
> I think we're g2g after jimb's comment is addressed.

this is fixed

cwfitzgerald reviewed (approved):
Thanks!

cwfitzgerald commented:
Got a conflict on `image.spvasm` - if you merge trunk then regenerate the files, this should resolve the conflict. Then ping me

ygdrasil-io committed:
regenerate image.spvasm
